src/lib/util/vec.ml

```diff
@@ -9,144 +9,144 @@
     done;
-  t.sz <- t.sz - i
-
-let pop t = assert (t.sz >=1); t.sz <- t.sz - 1
-
-let size t = t.sz
-
-let is_empty t = t.sz = 0
-
-let grow_to t new_capa =
-  let data = t.data in
-  let capa = Array.length data in
-  if new_capa > capa then
-    t.data <-
-      Array.init new_capa
-        (fun i -> if i < capa then data.(i) else t.dummy)
-
-let grow_to_double_size t =
-  let n = max 1 (Array.length t.data) in grow_to t (2 * n)
-
-let grow_to_by_double t new_capa =
-  let new_capa = max 1 new_capa in
-  let data = t.data in
-  let capa = ref (max 1 (Array.length data)) in
-  while !capa < new_capa do capa := 2 * !capa done;
-  grow_to t !capa
-
-
-let is_full t = Array.length t.data = t.sz
-
-let push t e =
-  (* Printer.print_dbg
-     "push; sz = %d et capa=%d@." t.sz (Array.length t.data);*)
-  if is_full t then grow_to_double_size t;
-  t.data.(t.sz) <- e;
-  t.sz <- t.sz + 1
-
-let push_none t =
-  if is_full t then grow_to_double_size t;
-  t.data.(t.sz) <- t.dummy;
-  t.sz <- t.sz + 1
+    false
+  with Terminate -> true
 
-let last t =
-  let e = t.data.(t.sz - 1) in
-  assert (not (e == t.dummy));
-  e
+let for_all p vec = not @@ exists (fun x -> not @@ p x) vec
 
-let get t i =
-  assert (i < t.sz);
-  let e = t.data.(i) in
-  if e == t.dummy then raise Not_found
-  else e
-
-let set t i v =
-  t.data.(i) <- v;
-  t.sz <- max t.sz (i + 1)
-
-let set_size t sz = t.sz <- sz
-
-let copy t =
-  let data = t.data in
-  let len = Array.length data in
-  let data = Array.init len (fun i -> data.(i)) in
-  { data=data; sz=t.sz; dummy = t.dummy }
-
-let move_to t t' =
-  let data = t.data in
-  let len = Array.length data in
-  let data = Array.init len (fun i -> data.(i)) in
-  t'.data <- data;
-  t'.sz <- t.sz
-
-
-let remove t e =
-  let j = ref 0 in
-  while (!j < t.sz && not (t.data.(!j) == e)) do incr j done;
-  assert (!j < t.sz);
-  for i = !j to t.sz - 2 do t.data.(i) <- t.data.(i+1) done;
-  pop t
-
-
-let fast_remove t e =
-  let j = ref 0 in
-  while (!j < t.sz && not (t.data.(!j) == e)) do incr j done;
-  assert (!j < t.sz);
-  t.data.(!j) <- last t;
-  pop t
-
-
-let sort t f =
-  let sub_arr = Array.sub t.data 0 t.sz in
-  Array.fast_sort f sub_arr;
-  t.data <- sub_arr
-
-let iter vec f =
+let fold f acc vec =
+  let acc = ref acc in
   for i = 0 to size vec - 1 do
-    f (get vec i)
-  done
-(*
-template<class V, class T>
-static inline void remove(V& ts, const T& t)
-{
-    int j = 0;
-    for (; j < ts.size() && ts[j] != t; j++);
-    assert(j < ts.size());
-    ts[j] = ts.last();
-    ts.pop();
-}
-#endif
-
-template<class V, class T>
-static inline bool find(V& ts, const T& t)
-{
-    int j = 0;
-    for (; j < ts.size() && ts[j] != t; j++);
-    return j < ts.size();
-}
-
-#endif
-*)
+    let elt = Array.unsafe_get vec.data i in
+    if not (elt == vec.dummy) then
+      acc := f !acc elt
+  done;
+  !acc
+
+let to_list a = Array.to_seq a.data |> List.of_seq
```

Comment:
I'd expect this to return a list of length [a.sz], not [Array.length a.data].

Reply:
You are right. I will fix it.